src/LinearAlgebra/SparseMatrix/CsrMatrix.hpp

```diff
@@ -0,0 +1,172 @@
+                                  memory::MemorySpace memspace);
+
+      int destroyMatrixData(memory::MemorySpace memspace);
+
+      virtual void print(std::ostream& file_out, index_type indexing_base) = 0;
+
+      virtual int syncData(memory::MemorySpace memspaceOut) = 0;
+
+      // update Values just updates values; it allocates if necessary.
+      // values have the same dimensions between different formats
+      virtual int copyValues(const real_type*    new_vals,
+                             memory::MemorySpace memspaceIn,
+                             memory::MemorySpace memspaceOut);
+
+      // set new values just sets the pointer, use caution.
+      virtual int setValuesPointer(real_type*          new_vals,
+                                   memory::MemorySpace memspace);
+
+    protected:
+      SparseFormat sparse_format_{NONE}; ///< Matrix format
+      index_type   n_{0};                ///< number of rows
+      index_type   m_{0};                ///< number of columns
+      index_type   nnz_{0};              ///< number of non-zeros
+
+      bool is_symmetric_{false}; ///< symmetry flag
+      bool is_expanded_{false};  ///< "expanded" flag
+
+      // host data
+      index_type* h_row_data_{nullptr};   ///< row data (HOST)
+      index_type* h_col_data_{nullptr};   ///< column data (HOST)
+      real_type*  h_val_data_{nullptr};   ///< value data (HOST)
+      bool        h_data_updated_{false}; ///< HOST update flag
+
+      // gpu data
+      index_type* d_row_data_{nullptr};   ///< row data (DEVICE)
+      index_type* d_col_data_{nullptr};   ///< column data (DEVICE)
+      real_type*  d_val_data_{nullptr};   ///< value data (DEVICE)
+      bool        d_data_updated_{false}; ///< DEVICE update flag
+
+      void setNotUpdated();
+
+      // Data ownership flags
+      bool owns_cpu_sparsity_pattern_{false}; ///< for row/col data
+      bool owns_cpu_values_{false};           ///< for nonzero values
+
+      bool owns_gpu_sparsity_pattern_{false}; ///< for row/col data
+      bool owns_gpu_values_{false};           ///< for nonzero values
+
+      MemoryHandler mem_; ///< Device memory manager object
+    };
+
+    class CsrMatrix : public Sparse
+    {
+    public:
+      CsrMatrix();
+
+      CsrMatrix(index_type n, index_type m, index_type nnz);
+
+      CsrMatrix(index_type n,
+                index_type m,
+                index_type nnz,
+                bool       symmetric,
+                bool       expanded);
+
+      CsrMatrix(index_type          n,
+                index_type          m,
+                index_type          nnz,
+                bool                symmetric,
+                bool                expanded,
+                index_type**        rows,
+                index_type**        cols,
+                real_type**         vals,
+                memory::MemorySpace memspaceSrc,
+                memory::MemorySpace memspaceDst);
```

Comment:
```suggestion
                memory::MemorySpace memspaceSrc = memory::HOST,
                memory::MemorySpace memspaceDst = memory::HOST);
```
Let memory spaces default to `HOST` in each function. In that way, GridKit user would not need to specify memory spaces when working on a CPU hardware (which is most of the use cases).